Error handling -> Targets Issue Better Error Handling #13

MO-W59 opened the pull request:
### Implements

Added `cmd_err_handler.py` in `./cogs` overriding discord.py's `on_command_error()`.
Added new cog to bot.

Removed `if not self.game:` checks from `game_commands.py`

#### Functionality

Here's what I have for some common errors right now. Message format can be changed if need be.

![Screenshot 2024-01-22 095245](https://github.com/Swiddis/word-debt-bot/assets/134148713/dbd734b8-d698-459a-bec9-757cd521eb4c)

I was still able to get the 'game not loaded' message by forcing the game state to none. (More on this in Unit Test Update)

![Screenshot 2024-01-22 113358](https://github.com/Swiddis/word-debt-bot/assets/134148713/b7edcd52-5aab-46d2-989b-10eb8c36f41e)

### Further actions

#### Handle Base Error Case with log?

I was thinking it might be a good idea to still pass unhandled errors to the log, such as ExtensionErrors or others that may not be encountered as of now. `elif` branches could then be added to handle them as the project grows in scope.

However, I'm having trouble on accessing the log file, currently the base cases just send the error and type to the user. Will look back into it if logging unknown edge cases is something we want.

#### Add missing Error Types

There's a long list of error types for CommandErrors, I think I've gotten the most important ones, I'll go back through the [list](https://discordpy.readthedocs.io/en/latest/ext/commands/api.html#exception-hierarchy) but feel free to point out any big ones I've missed.

#### Update Unit Tests

The unit tests for checking game state are currently failing due to the removal of the `if not self.game:` checks, I am trying to make replacements but its proving difficult. Still working on it. 

Let me know what you think,
-Greg

Swiddis commented on src/word_debt_bot/cogs/cmd_err_handler.py:
```diff
@@ -0,0 +1,66 @@
+        elif isinstance(err.__cause__, AttributeError) and not self.game:
+            await ctx.send("Game not loaded. (Yell at Toast!)")
+
+        elif isinstance(err, commands.CommandInvokeError):
```
Is this error one that wraps around all exceptions thrown by commands internally? Wondering if we can't also throw in handling for `ValueError` (generally because arguments were invalid) and `KeyError` (often if a user isn't registered).

MO-W59 replied:
From my understanding of `CommandInvokeError` that is correct. "Exception raised when the command being invoked raised an exception." Has attribute `original` that can be retrieved with `__cause__` which is an `Exception`.

Removing the `try` `except` from the `log` command (like below) is then handled by the `CommandInvokeError` in `cmd_err_handler.py`.

```python
    @commands.command(name="log")
    async def log(self, ctx, words: int):
        new_debt = self.game.submit_words(str(ctx.author.id), words)
        self.journal({"command": "log", "words": words, "user": str(ctx.authorid)})
        await ctx.send(f"Logged {words:,} words! New debt: {new_debt:,}")
```

So I'm confident we could migrate the `try` `except` blocks in `game_commands.py` if we wish, and even go so far as basing the `if` logic off of `ctx.command` as well for more accurate error messages. I like this idea too because it moves error handling all to one area rather than splitting between two files and cleans up the game command methods a little bit.

-> Added to TODOs

MO-W59 replied:
#### Update:
1. Migrated the `try` `except` blocks from `game_commands.py` to the error handler.
2. Added remaining 'expected' error types to handler.
3. Updated unit tests related to error handling.

#### Further Actions:
The unit tests involving the `log` command are currently only ensuring that the command is raising the exception specified. I have yet to find a replacement for the `invoke()` method in these situations since it cannot take arguments other than `ctx`, I've posted a question on [StackOverflow](https://stackoverflow.com/questions/77883736/how-to-ensure-discord-bot-is-sending-message-from-on-command-error-when-a-comm) see if anything turns up.

However, the bot currently responds appropriately to such errors for example:
![Screenshot 2024-01-27 101110](https://github.com/Swiddis/word-debt-bot/assets/134148713/499e28c8-8b68-4c90-84b9-32e67d14b65f)

MO-W59 replied:
#### Update:

Tests are working now, I wrapped them in another command to invoke to pass in the parameters. 

Now investigating some pytest warnings for `RuntimeWarning: coroutine 'AsyncMockMixin._execute_mock_call' was never awaited`

```
tests/test_game_commands_cog.py::test_submit_words_with_error
tests/test_game_commands_cog.py::test_submit_words_with_no_register
tests/test_game_commands_cog.py::test_submit_words_no_game
tests/test_game_commands_cog.py::test_request_leaderboard_no_game
  C:\Users\gregn\AppData\Local\pypoetry\Cache\virtualenvs\word-debt-bot-yFc_9xiU-py3.11\Lib\site-packages\discord\ext\commands\core.py:656: RuntimeWarning: coroutine 'AsyncMockMixin._execute_mock_call' was never awaited
```

Swiddis commented on src/word_debt_bot/cogs/cmd_err_handler.py:
```diff
@@ -0,0 +1,66 @@
+        if hasattr(ctx.command, "on_error"):
+            return
+
+        if isinstance(err, commands.CommandNotFound):
```
I'm inclined to just ignore this error, I find it annoying when bots complain about invalid commands when I wasn't trying to use them in the first place.

MO-W59 replied:
Removed from local working version.

Swiddis reviewed (commented):
I like the direction so far, aside from the tests and open TODOs it looks good to go. I have some minor points on specific error behavior.

Swiddis commented on src/word_debt_bot/cogs/cmd_err_handler.py:
```diff
@@ -0,0 +1,92 @@
+class CmdErrHandler(commands.Cog, name="Command Error Handler"):
+    def __init__(self, bot: client.WordDebtBot, game: game.WordDebtGame):
+        self.bot = bot
+        self.game = game
```
Does the error handler need a game field? I don't see it used anywhere

Swiddis replied:
Ah, found it

Swiddis reviewed (approved):
LGTM! :rocket:

The workaround with the error handler for tests is a little messy but not a blocking issue, I can maybe mess with it another time. There's probably a clever way to turn it into a fixture. Maybe some integration tests would be more clear additions too.

Ready to merge whenever you think the PR is ready

MO-W59 commented:
### Update:

- Merged commits from Dev -> new `buy` & `info` command etc.
- Migrated `buy` & `info` command errors to error handler.
- Added `ExpectedClosingQuoteError` to handler
- Updated `test_buy_with_value_error` to use workaround like other unit tests needing an invoke error.

Bot looks to be handling buy errors appropriately in testing after the migration to the error handler and all unit tests are passing. I'll label the PR ready for review, I'm not the biggest fan of the unit test workaround either, if anything shows up on the SO question I'll let you know.

#### Pytest Warnings:
The pytest warning for a few of the unit tests;

 `RuntimeWarning: coroutine 'AsyncMockMixin._execute_mock_call' was never awaited`

is listing `view.skip_ws()` as the function not awaited. This function is in `discord.ext.commands.core`. I went in and added await to the function and the warnings went away (however, a different unit test failed). I'm thinking the function's purpose is to skip over whitespaces given to commands. It would be ignorant of me to blame the discord library for the warnings, but it seems like it could be a possibility.
